### PR TITLE
Introducing new team member Quark Script Agent

### PR DESCRIPTION
**Pull request by pulorsok**
# Quark Script Agent

Introducing Quark's new member, the Quark Script Agent, the first AI assistant in the Quark team. This agent enables users to perform analyses using natural language, without the need for programming or scripting expertise, making the process simple and user-friendly.

The Quark Script Agent integrates with LangChain, which utilizes OpenAI's large language models to act as a bridge between natural language and the Quark Script API. LangChain defines the Quark Script API as a tool that large language models can understand and use. This means that users can easily call new analysis APIs using natural language commands by simply adding new tools as needed.

## Showcase: Detecting CWE-798 with Quark Script Agent
Here's an example of using the Quark Script Agent with the `quarkScriptAgent.py`. This agent can currently detect [CWE-798](https://cwe.mitre.org/data/definitions/798.html) vulnerability in the [ovaa.apk](https://github.com/oversecured/ovaa). See the details below.

### Quick Start

1. clone the repository:
```
git clone https://github.com/quark-engine/quark-script.git
```

2. Install the required packages:
```
pip install -r requirements.txt
```

3. Run the script:
```
python quarkScriptAgent.py
```

4. Result:

截圖 2024-07-26 下午3 39 12

### Decode the Prompts
Here are two prompts, each for executing different analysis processes.

```
1st Prompt: Initialize the rule instance with the rule path set to "rule.json"
```
Used Quark Script APIs/Tools that LLM used:  `Rule()`

```
2nd Prompt: Run Quark Analysis using the rule instance on the apk sample "ovaa.apk", 
            and Check if the parameters are hard-coded. If yes, display the hard-coded values.
```
Used Quark Script APIs/Tools that LLM used: `runQuarkAnalysis()`, `getParameterValues()` and `isHardCoded()`

The `Rule()`, `runQuarkAnalysis()`, `getParameterValues()`, and `isHardCoded()` functions are treated as **tools** within LangChain, enabling them to be invoked through the `gpt-4o` model to analyze and identify [CWE-798](https://cwe.mitre.org/data/definitions/798.html) vulnerabilities in the [ovaa.apk](https://github.com/oversecured/ovaa) sample.

截圖 2024-07-26 下午9 25 23

* Notes: 
  1. Since LangChain currently does not support passing Python instances between tools, we are temporarily using global variables to pass parameters between tools in `quarkScriptAgent.py`.
  2. Place the rules, samples, and `quarkScriptAgent.py` in the same folder; the LLM will automatically find files with matching names.
  3. A web GUI is under construction, please stay tuned!

### CODE REVIEW

**Comment by 18z** [comment]
# Quark Script Agent

Introducing Quark's new member, the Quark Script Agent, the first AI assistant in the Quark team. This agent enables users to perform analyses using natural language, without the need for programming or scripting expertise, making the process simple and user-friendly.

The Quark Script Agent integrates with LangChain, which utilizes OpenAI's large language models to act as a bridge between natural language and the Quark Script API. LangChain defines the Quark Script API as a tool that large language models can understand and use. This means that users can easily call new analysis APIs using natural language commands by simply adding new tools as needed.

## Showcase: Detecting CWE-798 with Quark Script Agent
Here's an example of using the Quark Script Agent with the `quarkScriptAgent.py`. This agent can currently detect [CWE-798](https://cwe.mitre.org/data/definitions/798.html) vulnerability in the [ovaa.apk](https://github.com/oversecured/ovaa). See the details below.

### Installation

To install the Quark Script Agent, you need to install the following dependencies:

1. clone the repository:
```
git clone https://github.com/quark-engine/quark-script.git
```

2. Install the required packages:
```
pip install -r requirements.txt
```

3. Run the script:
```
python quarkScriptAgent.py
```

### Result

截圖 2024-07-26 下午3 39 12

### Decode the Prompts
Here are two prompts, each for executing different analysis processes.

```
1st Prompt: Initialize the rule instance with the rule path set to "rule.json"
```
Used Quark Script APIs/Tools that LLM used:  `Rule()`

```
2nd Prompt: Run Quark Analysis using the rule instance on the apk sample "ovaa.apk", 
            and Check if the parameters are hard-coded. If yes, display the hard-coded values.
```
Used Quark Script APIs/Tools that LLM used: `runQuarkAnalysis()`, `getParameterValues()` and `isHardCoded()`


### Architecture
The `Rule()`, `runQuarkAnalysis()`, `getParameterValues()`, and `isHardCoded()` functions are treated as **tools** within LangChain, enabling them to be invoked through the `gpt-4o` model to analyze and identify [CWE-798](https://cwe.mitre.org/data/definitions/798.html) vulnerabilities in the [ovaa.apk](https://github.com/oversecured/ovaa) sample.

截圖 2024-07-26 下午3 40 32


* Notes: 
  1. Since LangChain currently does not support passing Python instances between tools, we are temporarily using global variables to pass parameters between tools in `quarkScriptAgent.py`.
  2. This is a POC (Prove of Concept) the paths of the apk file and the rule file are hard-coded in the quarkScriptAgent.py
  3. A web GUI is under construction, please stay tuned!

File: quarkScriptAgent.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    api_key = input("OpenAI API Key: ")
+    os.environ["OPENAI_API_KEY"] = api_key
+
+
+@tool
+def defineRuleByJson():
+    """
+    This instance defines the detection rule
+    by loading the rule from a JSON.
+    """
```
Comment (haeter525) [inline quarkScriptAgent.py:26]: This tool only loads the rule from a given path. We could keep its name and its docstring simple. For example,

```python
@tool
def loadRule(rule_path: str):
    """
    Given a rule path,
    this instance loads a rule from the rule path.
    """
    ...
```

File: quarkScriptAgent.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    ruleInstance = Rule("rule.json")
+
+    return "Rule defined successfully"
+
+
+@tool
+def runQuarkAnalysis(samplePath):
+    """
+    Given detection rule and target sample,
+    this instance runs the basic Quark.
```
Comment (haeter525) [inline quarkScriptAgent.py:38]: I suggest replacing `the basic Quark` with `the Quark analysis` to make the docstring consistent with the tool's name.

File: quarkScriptAgent.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    global quarkResultInstance
+
+    quark = _getQuark(samplePath)
+    quarkResultInstance = QuarkResult(quark, ruleInstance)
+
+    return "Quark analysis completed successfully"
+
+
+@tool
+def getBehaviorOccurLise():
```
Comment (haeter525) [inline quarkScriptAgent.py:51]: There is a typo, `Lise`.

File: quarkScriptAgent.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+def isHardCoded():
+    """
+    Given the parameter values,
+    this instance checks if the parameter values are hard-coded.
+    """
+
+    global parameters
+    global quarkResultInstance
+
+    # check parameters values are hard-coded
```
Comment (haeter525) [inline quarkScriptAgent.py:92]: There is also a typo, `parameter values`.

File: quarkScriptAgent.py
Comment (haeter525) [inline quarkScriptAgent.py:1]: I recommend formatting the code with `black` to provide a consistent coding style.

**Review by haeter525 — CHANGES_REQUESTED** [review]

File: quarkScriptAgent.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+    # check parameter values are hard-coded
+    if parameters in quarkResultInstance.getAllStrings():
+        return parameters
+
+    return False
+
+
+tools = [
+    defineRuleByJson,
```
Comment (haeter525) [inline quarkScriptAgent.py:100]: This should be `loadRule`.

File: quarkScriptAgent.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    runQuarkAnalysis,
+    getBehaviorOccurLise,
```
Comment (haeter525) [inline quarkScriptAgent.py:102]: This should be `getBehaviorOccurList`.

File: quarkScriptAgent.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+from quark.script import Rule, _getQuark, QuarkResult
+
+if "OPENAI_API_KEY" not in os.environ:
+    api_key = input("OpenAI API Key: ")
+    os.environ["OPENAI_API_KEY"] = api_key
+
+
+@tool
+def loadRule(rulePath):
```
Comment (haeter525) [inline quarkScriptAgent.py:22]: Please add a type hint for `rulePath`.
```python
def loadRule(rulePath: str):
```


Without the type hint, the LLM can not pass the rule path to this tool. Here is the error output.

```
User Input: Initialize the rule instance with the rule path set to "rule.json"

Agent:  The rule has been successfully initialized with the path set to "rule.json". What would you like to do next?

User Input: Run Quark Analysis using the rule instance on the apk sample "ovaa.apk",             and Check if the parameters are hard-coded. If yes, display the hard-coded values.
Traceback (most recent call last):
  File "/Users/shengfeng/codespace/quark-script/quarkScriptAgent.py", line 137, in <module>
    response = agent_executor.invoke({
  File "/Users/shengfeng/codespace/quark-script/.venv/lib/python3.9/site-packages/langchain/chains/base.py", line 166, in invoke
    raise e
  File "/Users/shengfeng/codespace/quark-script/.venv/lib/python3.9/site-packages/langchain/chains/base.py", line 156, in invoke
    self._call(inputs, run_manager=run_manager)
  File "/Users/shengfeng/codespace/quark-script/.venv/lib/python3.9/site-packages/langchain/agents/agent.py", line 1612, in _call
    next_step_output = self._take_next_step(
  File "/Users/shengfeng/codespace/quark-script/.venv/lib/python3.9/site-packages/langchain/agents/agent.py", line 1318, in _take_next_step
    [
  File "/Users/shengfeng/codespace/quark-script/.venv/lib/python3.9/site-packages/langchain/agents/agent.py", line 1318, in <listcomp>
    [
  File "/Users/shengfeng/codespace/quark-script/.venv/lib/python3.9/site-packages/langchain/agents/agent.py", line 1403, in _iter_next_step
    yield self._perform_agent_action(
  File "/Users/shengfeng/codespace/quark-script/.venv/lib/python3.9/site-packages/langchain/agents/agent.py", line 1425, in _perform_agent_action
    observation = tool.run(
  File "/Users/shengfeng/codespace/quark-script/.venv/lib/python3.9/site-packages/langchain_core/tools.py", line 634, in run
    raise error_to_raise
  File "/Users/shengfeng/codespace/quark-script/.venv/lib/python3.9/site-packages/langchain_core/tools.py", line 607, in run
    response = context.run(self._run, *tool_args, **tool_kwargs)
  File "/Users/shengfeng/codespace/quark-script/.venv/lib/python3.9/site-packages/langchain_core/tools.py", line 944, in _run
    return self.func(*args, **kwargs)
  File "/Users/shengfeng/codespace/quark-script/quarkScriptAgent.py", line 29, in loadRule
    ruleInstance = Rule(rulePath)
  File "/Users/shengfeng/codespace/quark-script/.venv/lib/python3.9/site-packages/quark/core/struct/ruleobject.py", line 38, in __init__
    with open(ruleJson) as json_file:
FileNotFoundError: [Errno 2] No such file or directory: 'path/to/rule'
```

**Review by haeter525 — CHANGES_REQUESTED** [review]

**Review by haeter525 — APPROVED** [review]
LGTM. Thank @pulorsok !

**Pull request merged**